### DISCUSSION TITLE
Is it possible to save plots as image

### DISCUSSION THREAD

**Discussion by mrkaragoz** [emilk/egui#1742] [Q&A]
I would like to save `egui::plot`' s as images. Is it possible with the current  `egui`? If not, what should be done to achieve this? What part of the code should I look for this feature to add? Thanks in advance.

**Comment by AtmoPierce** [comment]
While doing some digging... 

Maybe here: https://github.com/emilk/egui/blob/0.18.1/epaint/src/image.rs

Would love to see this!

**Comment by Hidan0** [comment]
If the plot takes up the whole screen you could save the entire frame.

You can archive this by sending a `frame.request_screenshot()` from `App::update()` and then retrieving the screenshot into `App::post_rendering()` using `frame.screenshot()`.

For instance: 
```rust 
impl eframe::App for App {
    fn update(&mut self, ctx: &Context, frame: &mut eframe::Frame) {
        ...
        frame.request_screenshot();
        ...
    }

    fn post_rendering(&mut self, _window_size_px: [u32; 2], frame: &eframe::Frame) {
       ...
       if let Some(screenshot) = frame.screenshot() {
           self.save_current_plot_to_png(&screenshot);
       }
       ...
    }
}
```
[Here](https://github.com/Hidan0/value-plotter/blob/d45aa9e54e316c5731c933cba918610c8a33ab37/src/app.rs#L61) you can find how i used it.

**Reply by stonerfish** [reply]
If you know the location of the rectangle that contains the the plot, you may be able to extract that rectangle from the screenshot following this example.

https://docs.rs/eframe/latest/eframe/struct.Frame.html#example-generating-a-capture-of-everything-within-a-square-of-100-pixels-located-at-the-top-left-of-the-app-and-saving-it-with-the-image-crate

**Reply by jsalort** [reply]
I have a similar need as OP, and I tried the method described in the link, but I get:
```
 no method named `request_screenshot` found for mutable reference `&mut eframe::Frame` in the current scope
```

And I can't find either `request_screenshot` or `screenshot` in eframe API documentation.
Do I need to enable some feature ?
The example of the `value-plotter` project does not seem to be using any additionnal feature in its Cargo.toml...

**Comment by wangxiaochuTHU** [comment]
> I have a similar need as OP, and I tried the method described in the link, but I get:
> 
> ```
>  no method named `request_screenshot` found for mutable reference `&mut eframe::Frame` in the current scope
> ```
> 
> And I can't find either `request_screenshot` or `screenshot` in eframe API documentation. Do I need to enable some feature ? The example of the `value-plotter` project does not seem to be using any additionnal feature in its Cargo.toml...

`eframe` has the `__screenshot` feature.

**Comment by jsalort** [comment]
Thanks for you reply @wangxiaochuTHU. After digging into the examples, I realized that there is an up-to-date example here,
[screenshot](https://github.com/emilk/egui/tree/master/examples/screenshot).

I've tried that method, and it works, e.g. `ctx.send_viewport_cmd(egui::ViewportCommand::Screenshot);` to ask for the screenshot, then detect the `egui::Event::Screenshot` event with its associated image, and get the data from this image with its `as_raw()` method.
